### ISSUE TITLE
Auto-attach: special test code needed

### ISSUE THREAD

**Issue by CharlieFenton** [BOINC/boinc#2241]
To test the new auto-attach code (PR #2207), special temporary or configurable code needs to be added to the server and client:

* The _account_data.txt_ file downloaded by the test project now has a project ID of 100, but there is no project with this ID in the _all_projects_list.xml_ file. The test clients will need to use a test version of the _all_projects_list.xml_ file which includes the test project, rather than the standard _all_projects_list.xml_ file.

* The test project downloads the current release build, which does not have any of the auto-attach code. The test project server needs to download the test builds of the client installers instead of the current release builds.

It would probably be best to use a #define whose value can be modified to enable or disable the test code so that it can be re-enabled for use in future testing as well as being used for initial testing.

**Comment by TheAspens** [comment]
Based on the discussion in the commit, it sounds like this issue means two things should be changed just before the release branch is cut:

1. Remove the test project from the all_projects.xml list included in the distribution
2. Change the overwrite setting on the installer so that it will not replace an existing file (check this setting for both the windows and mac installers)

I'm tagging this issue in the client 7.10 release project so that it is seen for that.  

Please confirm that the above correctly reflects the agreement on this issue.

**Comment by CharlieFenton** [comment]
As I wrote in my initial comment for the issue, I _very_ strongly disagree with this approach. It will require that I modify the Mac installer to overwrite an existing _all_projects_list.xml_ file, and then change it again for final release, and we will also need to do the same for the client code to prevent it from downloading a standard _all_projects_list.xml_ file during testing. Similarly, it requires change it twice in the future every time we ned to test the auto-attach with a test project, which I expect will happen from time to time. 

Sorry, I am not willing to do make that change in the Mac installer. I have proposed a different method in this PR, and am not happy that my concerns and proposal are being ignored.

**Comment by CharlieFenton** [comment]
> Please confirm that the above correctly reflects the agreement on this issue.

Sorry, no. I strongly believe that a branch with temporary test code in it should never be merged into master, nor should code with open issues that require making a change later. I feel that our goal should be that all code in master could be released for alpha testing at any time with no further changes. That is definitely not the case here. I see no reason why this branch needed to be merged prematurely. All testing could have been done within this branch.

**Comment by TheAspens** [comment]
Charlie - my apologies.  I clearly did not understand your proposal.  There was no intent to ignore your concerns .  

I took your statement from the pull request "I have opened issues #2240, #2241 and #2242. I will leave it up to @TheAspens to decide whether these need to be resolved before merging." to indicate that you were ok with a merge and that open issues could be addressed with subsequent pull requests.

I apologize that I did not understand your subsequent statements to be a "veto" (in the vocabulary of the consensus voting).

In order to address this I propose that we create two branches now from master:
- dpa_autoattach
- dpa_revert_autoattach_test

dpa_revert_autoattach_test will contain the changes necessary to remove any test code and as soon as that is complete we will merge to master.

dpa_autoattach will still have that test logic and david can do all of his testing from there as well as all additional work on the feature.  Once he is done and resolved issues he can create a pull request to merge.

Is this a reasonable approach to solve this situation?

**Comment by TheAspens** [comment]
Also - can you please chime into the discussion on this thread: https://groups.google.com/forum/#!topic/boinc_admin/bfRl6-heSYs as it speaks to this exact issue and the community needs to determine the standard.

**Comment by CharlieFenton** [comment]
> Is this a reasonable approach to solve this situation?

That still requires that I modify the Mac installer code to overwrite the _all_projects_list.xml_ file, echoing what was done in the Windows installer, which I do not want to do because I feel it is the wrong approach. I prefer the method I outlined in my comment in https://github.com/BOINC/boinc/pull/2207#commitcomment-25626062; I have not seen any reference to that in this discussion; it seems that the approach modifying the Windows installer has been accepted despite my objections.

Since this has already been merged, let's not complicate matters further by creating yet more branches and reverting. Here is what I would like to see happen:

* Revert the change to the Windows installer that allows it to overwrite the _all_projects_list.xml_ file
* Put the test _all_projects_list.xml_ file in the test project server at a special URL 
* add a `#define TEST_AUTO-ATTACH 1` to the client
* Add code to the client that downloads the _all_projects_list.xml_ file at the start of an auto attach (or whenever a new version has being installed):
(a) from the standard URL if #define TEST_AUTO-ATTACH is 0 (false)
(b) from the special test project server URL if #define TEST_AUTO-ATTACH is 1 (true)

The current approach with the installer overwriting the _all_projects_list.xml_ file will fail once the code recommended in #2242 (Auto-attach: client must get latest all_projects.xml list) is added. I feel that #2242 should be addressed before we test this; otherwise we are not testing the full implementation. The steps above address issue #2242.

And as testing proceeds, we may find additional things that need to be different for testing vs production. If that happens, the #if TEST_AUTO-ATTACH can easily be used in those cases as well.

**Comment by TheAspens** [comment]
This makes sense as it eliminates the need for any changes to the installer as you have mentioned.

What do you think would be preferable.  A cc_config.xml option to define the location of the all_projects_list.xml or a compile option you have detailed?  By making it a cc_config.xml option it doesn't require a recompile which will make testing easier and would let non-technical people test with it.  On the flip side it requires putting the cc_config.xml file in the appropriate location before the installation occurs which would slightly change the initial installation test case.

**Comment by TheAspens** [comment]
Actually a cc_config option to specify this already exists: https://boinc.berkeley.edu/wiki/Client_configuration : client_version_check_url

So the question is - is this sufficient?  

I agree that #2242 must also be implemented

**Comment by CharlieFenton** [comment]
We need to test both with a virgin installation and an existing installation. A virgin installation does not have a BOINC Data directory, so there is nowhere to put the cc_config file.

**Comment by CharlieFenton** [comment]
Thank you for finally considering my suggested changes to testing and looking at alternatives. One additional disadvantage of the current approach (overwriting the _all_projects_list.xml_ file in the installer) is that it is not cross-platform. The installer is one of the few components (perhaps the only one) that is different for each platform. In fact, there typically is no installer for Linux, so that approach can't be used for testing on Linux. A cross-platform approach (such as the one I have proposed) is fully cross-platfrom..

**Comment by CharlieFenton** [comment]
> Revert the change to the Windows installer that allows it to overwrite the all_projects_list.xml file
> Put the test all_projects_list.xml file in the test project server at a special URL
> add a #define TEST_AUTO-ATTACH 1 to the client
> Add code to the client that downloads the all_projects_list.xml file at the start of an auto attach (or whenever a new version has being installed):
> (a) from the standard URL if #define TEST_AUTO-ATTACH is 0 (false)
> (b) from the special test project server URL if #define TEST_AUTO-ATTACH is 1 (true)

Also:
* if #define TEST_AUTO-ATTACH is 1 (true), then write a message to the Event Log in red saying something like "WARNING: this client is configured in Auto-Attach test mode.For testing only. Do not release."

**Comment by CharlieFenton** [comment]
I do like the idea of having something external to the code to trigger Auto-Attach test mode, rather than having to recompile code or rebuild the installer. One thought is to require the tester to create a BOINC Data directory before doing a "virgin" test install, but putting a file named "test_auto-install" in that otherwise empty directory. To test installing over an existing installation, the tester would add that file to the existing BOINC Data directory.

The disadvantage is that the built client would always include the extra test code even though it would normally not be executed.

But I'll think about this some more.

**Comment by CharlieFenton** [comment]
> https://boinc.berkeley.edu/wiki/Client_configuration : client_version_check_url

That changes where the client checks for a new version of the _client_ installer, We need to check for a new version of the _all_projects_list.xml_ file, which lists where to find _projects_.

**Comment by TheAspens** [comment]
Shoot - you are right.  Sorry for the misdirect. 

But let me know your thoughts after having thought on it some more.  We could add the feature. to the cc_config.xml if needed.

I like the idea of being able to take a released version of the client and being able to test it without having to recompile it - that is probably my biggest reason for going with.

**Comment by TheAspens** [comment]
> One thought is to require the tester to create a BOINC Data directory before doing a "virgin" test install, but putting a file named "test_auto-install" in that otherwise empty directory. 

I think that this would be viable and reasonable for testers to do and would not invalidate the test case.

> To test installing over an existing installation, the tester would add that file to the existing BOINC Data directory.

I also agree.

>  The disadvantage is that the built client would always include the extra test code even though it would normally not be executed.

I think that adding code to production systems to enable testing or monitoring is acceptable and normal even if that code can be triggered by changing a flag.

**Comment by davidpanderson** [comment]
Once we get the server part in a vetted project (e.g. SETI@home) we can test with the standard project list. This should happen in a week or so.  We won't release it for testing before then.

**Comment by davidpanderson** [comment]
To clarify: there are 2 phases of testing for this feature.  The first one we're doing now; it requires a special project list.  The second one won't require this.

**Comment by davidpanderson** [comment]
Charlie: the Mac installer doesn't need to write the file.
Just make sure while testing that you delete the old one first.

**Comment by CharlieFenton** [comment]
> Charlie: the Mac installer doesn't need to write the file.
> Just make sure while testing that you delete the old one first.

OK, that will work as long as we are building from the branch, which contains the special test _all_projects_list.xml_ file. But that file should not be merged into master, so we will need to manually insert the special test _all_projects_list.xml_ file into the BOINC Data directory if we build from master.

**Comment by TheAspens** [comment]
I've opened PR #2251 to change the windows installer so that it does not overwrite the all_projects_list.xml during installation. 

Note that only the 64 bit windows installer was changed so only that is being reverted.  ( @davidpanderson  - please confirm this statement is true).

**Comment by davidpanderson** [comment]
I already did this in #2207

**Comment by TheAspens** [comment]
Since #2207 was merged it is closed.  I opened #2250 as the pr that continues tracking subsequent changes to the branch.

I see that change now - I missed it before.  I also created the stand alone PR so that it could be merged now to remove that from the code.  Feel free to close the PR if you and Charlie are fine with not merging the revert until the rest of the work is finished.
